### PR TITLE
use Intl APIs for filtering and sorting

### PR DESCRIPTION
**Pull request by joshuaellis**
### What does it do?

* Adds `useFilter` and `useCollator` hook into `@strapi/helper-plugin`
* replace `match-sorter` with a combination of these hooks to filter and sort in the following places:
	* Content-Manager nav bar
	* Content-Type-Builder nav bar
	* Settings users table
	* User & Permissions table

### Why is it needed?

* The CTB & CM changes were because the "search" was not working as expected e.g. you type `AD` and have a content-type `Event-Brand` it would still show incorrectly. There were also issues with localisation, some users were reporting things were incorrectly sorted in languages like german
* The tables were just hygiene to remove a dependency we no longer need.

### Related issue(s)/PR(s)

I could have sworn there were some issues reported, but it might have been enterprise customers.

* resolves CONTENT-1115

### To do

- [x] Docs
- [x] Tests

### CODE REVIEW

File: packages/core/admin/admin/src/content-manager/pages/App/LeftMenu/index.js [hunk 1]
```diff
@@ -16,51 +16,72 @@ import {
...
+  const { collectionTypeLinks, singleTypeLinks } = useSelector(modelLinksSelector, shallowEqual);
 
-  const toIntl = (links) =>
-    links.map((link) => {
-      return {
-        ...link,
-        title: formatMessage({ id: link.title, defaultMessage: link.title }),
-      };
-    });
+  const { contains } = useFilter(locale, {
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/pages/App/LeftMenu/index.js:30]: What do you think about `includes` instead to be inline with JS `String.includes`?

File: packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js [hunk 2]
```diff
@@ -111,12 +96,36 @@ const useContentTypeBuilderMenu = () => {
...
+        e.stopPropagation();
+
+        if (canOpenModalCreateCTorComponent) {
+          onOpenModalEditCategory(data.name);
+        } else {
+          toggleNotificationCannotCreateSchema();
+        }
+      },
+      links: components
+        .map((compo) => ({
```
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js:120]: 👀

File: packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js [hunk 3]
```diff
@@ -111,12 +96,36 @@ const useContentTypeBuilderMenu = () => {
...
+          name: compo.uid,
```
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js:121]: 👀

File: packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js [hunk 4]
```diff
@@ -111,12 +96,36 @@ const useContentTypeBuilderMenu = () => {
...
+          to: `/plugins/${pluginId}/component-categories/${category}/${compo.uid}`,
+          title: compo.schema.displayName,
```
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js:123]: 👀

File: packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js [hunk 5]
```diff
@@ -111,12 +96,36 @@ const useContentTypeBuilderMenu = () => {
...
       type: 'info',
       message: {
-        id: `${getTrad('notification.info.creating.notSaved')}`,
+        id: getTrad('notification.info.creating.notSaved'),
         defaultMessage: 'Please save your work before creating a new collection type or component',
       },
     });
   };
 
+  const componentsData = Object.entries(componentsGroupedByCategory)
```
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js:105]: Should these be memoized too?

File: packages/core/admin/admin/src/content-manager/pages/App/LeftMenu/index.js (hunk 1 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/pages/App/LeftMenu/index.js:30]: sure idm 👍🏼

File: packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js (hunk 5 above)
Comment (joshuaellis) [reply packages/core/content-type-builder/admin/src/components/ContentTypeBuilderNav/useContentTypeBuilderMenu.js:105]: They weren't before and in the interest of preservation i wanna keep it as close to how it was prior 😛

**Comment by joshuaellis** [comment]
@gu-stav i'm not going to add any tests for ListView because of the issues I spoke about in Slack, that setup should really be a separate PR which i'll follow up with (it's in JIRA).

File: docs/docs/core/helper-plugin/hooks/use-filter.mdx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+tags:
+  - hooks
+  - helper-plugin
+  - i18n
+---
+
+Provides localized string search functionality that is useful for filtering or matching items in a list. Options can be
+provided to adjust the sensitivity to case, diacritics, and other parameters.
+
+Stolen from [`react-aria`](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/i18n/src/useFilter.ts).
```
Comment (gu-stav) [inline docs/docs/core/helper-plugin/hooks/use-filter.mdx:13]: Nit: Since these are public-facing docs: should we say "borrow" instead? 🙈

**Review by gu-stav — COMMENTED** [review]
Works well on my side and the hooks look very useful to me! 👍🏼 

3 thoughts:

- we currently mix search patterns: for tables users have to hit "Enter" to submit the search where as for these lists we are using live-search. What do you think about unifying both (as a follow-up) to use "Enter"? Otherwise we maybe should talk with Maeva and see if we change the icon to a filter icon?
- I found it a bit unexpected searching for "R" returns "Address" as a first item because we use `includes`. Should we use `startsWith` instead? I think I would find that more intuitive.
- When searching in the CTB the component groups are not removed when they don't have any children, which gives me the impression as if they were matched. Should we not render them in case they do not contain any children?

Screenshot 2023-04-18 at 10 28 40

Good work! 🚀

File: docs/docs/core/helper-plugin/hooks/use-filter.mdx (hunk 6 above)
Comment (joshuaellis) [reply docs/docs/core/helper-plugin/hooks/use-filter.mdx:13]: Sure, if it makes you feel better lol

**Comment by gu-stav** [comment]
@joshuaellis sure - I don't think they need to be done as part of this and can be follow-ups.

**Review by gu-stav — APPROVED** [review]

**Comment by joshuaellis** [comment]
> @joshuaellis sure - I don't think they need to be done as part of this and can be follow-ups.

I think the last two can be done here for sure, the first one yes should be a follow up 👍🏼

**Comment by maevalienard** [comment]
Hello 👋 

> we currently mix search patterns: for tables users have to hit "Enter" to submit the search where as for these lists we are using live-search. What do you think about unifying both (as a follow-up) to use "Enter"? Otherwise we maybe should talk with Maeva and see if we change the icon to a filter icon?

Agree on unifying the methods. However I don't think hitting Enter is truly intuitive on the Internet anymore because many things seem to be live searched now. Regarding the filter icon, I think we should stick to the search one as this is the most common pattern users will look for, especially non-tech people. 
From what we understood during our research, non-tech users like to use search bars better than filters :) 

> I found it a bit unexpected searching for "R" returns "Address" as a first item because we use includes. Should we use startsWith instead? I think I would find that more intuitive.

+1, that's what we were mentioning with Paridhi in another US

> When searching in the CTB the component groups are not removed when they don't have any children, which gives me the impression as if they were matched. Should we not render them in case they do not contain any children?

+1

**Comment by joshuaellis** [comment]
Great, i'll make the `startsWith` change on the CM & CTB so the behaviour is identical & look at hiding the component categories if there are no children visible.

**Comment by gu-stav** [comment]
@joshuaellis Please don't forget to rebase on main because of https://github.com/strapi/strapi/pull/16387

**Review by gu-stav — APPROVED** [review]
Beautiful.

**Comment by gu-stav** [comment]
@joshuaellis Could you create a jira follow-up for the search pattern unification and reference this discussion so we don't forget about it? 🙏🏼

**Pull request merged**